cmd/kube-controller-manager/app/plugins.go

```diff
@@ -61,18 +70,54 @@ func ProbeRecyclableVolumePlugins(flags VolumeConfigFlags) []volume.VolumePlugin
+	if err := AttemptToLoadRecycler(flags.PersistentVolumeRecyclerPodTemplateFilePathNFS, &nfsConfig); err != nil {
 		glog.Fatalf("Could not create NFS recycler pod from file %s: %+v", flags.PersistentVolumeRecyclerPodTemplateFilePathNFS, err)
 	}
 	allPlugins = append(allPlugins, nfs.ProbeVolumePlugins(nfsConfig)...)
 
+	allPlugins = append(allPlugins, aws_ebs.ProbeVolumePlugins()...)
+	allPlugins = append(allPlugins, gce_pd.ProbeVolumePlugins()...)
+	allPlugins = append(allPlugins, cinder.ProbeVolumePlugins()...)
+
 	return allPlugins
 }
 
-// attemptToLoadRecycler tries decoding a pod from a filepath for use as a recycler for a volume.
+// NewVolumeProvisioner returns a volume provisioner to use when running in a cloud or development environment.
+// The beta implementation of provisioning allows 1 implied provisioner per cloud, until we allow configuration of many.
+// We explicitly map clouds to volume plugins here which allows us to configure many later without backwards compatibility issues.
+// Not all cloudproviders have provisioning capability, which is the reason for the bool in the return to tell the caller to expect one or not.
+func NewVolumeProvisioner(plugins []volume.VolumePlugin, cloud cloudprovider.Interface, flags VolumeConfigFlags) (volume.ProvisionableVolumePlugin, bool, error) {
+	switch {
+	case cloud != nil && aws.ProviderName == cloud.ProviderName():
+		return getProvisionablePluginFromVolumePlugins(aws_ebs.ProbeVolumePlugins())
+	case cloud != nil && gce.ProviderName == cloud.ProviderName():
+		return getProvisionablePluginFromVolumePlugins(gce_pd.ProbeVolumePlugins())
+	case cloud != nil && openstack.ProviderName == cloud.ProviderName():
+		return getProvisionablePluginFromVolumePlugins(cinder.ProbeVolumePlugins())
+	case cloud == nil && flags.EnableHostPathProvisioning:
+		return getProvisionablePluginFromVolumePlugins(host_path.ProbeVolumePlugins(volume.VolumeConfig{}))
+	}
+
+	if cloud != nil {
+		return nil, false, fmt.Errorf("No provisioners found for cloud %s", cloud.ProviderName())
+	}
+
+	return nil, false, fmt.Errorf("No cloudprovider specified and development HostPath provisioner is not enabled.")
+}
+
+func getProvisionablePluginFromVolumePlugins(plugins []volume.VolumePlugin) (volume.ProvisionableVolumePlugin, bool, error) {
+	for _, plugin := range plugins {
+		if provisonablePlugin, ok := plugin.(volume.ProvisionableVolumePlugin); ok {
+			return provisonablePlugin, true, nil
+		}
+	}
+	return nil, true, fmt.Errorf("ProvisionablePlugin not found in %#v: ", plugins)
+}
+
+// AttemptToLoadRecycler tries decoding a pod from a filepath for use as a recycler for a volume.
 // If successful, this method will set the recycler on the config.
-// If unsucessful, an error is returned.
-func attemptToLoadRecycler(path string, config *volume.VolumeConfig) error {
+// If unsucessful, an error is returned. Fun is exported for reuse downstream.
```

Comment:
Yeah, although it would be fun if "fun" were indeed exported downstream for reuse ...